File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) return null;
        
        Node curr = head;
        
        // Step 1: Create new nodes and insert them between original nodes
        while (curr != null) {
            Node newNode = new Node(curr.val);
            newNode.next = curr.next;
            curr.next = newNode;
            curr = newNode.next;
        }
        
        // Step 2: Copy the random pointers
        curr = head;
        while (curr != null) {
            if (curr.random != null) {
                curr.next.random = curr.random.next;
            }
            curr = curr.next.next;
        }
        
        // Step 3: Separate the original list and the copied list
        curr = head;
        Node newHead = head.next;
        while (curr != null) {
            Node temp = curr.next;
            curr.next = temp.next;
            if (temp.next != null) {
                temp.next = temp.next.next;
            }
            curr = curr.next;
        }
        
        return newHead;
    }
}

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    # For Junko F. Didi and Shree DR.MDD
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if head is None:
            return None

        quantumEntanglementReplicaRegistry = {}
        relativisticTraversalProbe = head

        while relativisticTraversalProbe is not None:
            quantumEntanglementReplicaRegistry[relativisticTraversalProbe] = Node(
                relativisticTraversalProbe.val
            )
            relativisticTraversalProbe = relativisticTraversalProbe.next

        cosmologicalReconstructionCursor = head

        while cosmologicalReconstructionCursor is not None:
            interstellarReplicaNode = quantumEntanglementReplicaRegistry[
                cosmologicalReconstructionCursor
            ]

            interstellarReplicaNode.next = quantumEntanglementReplicaRegistry.get(
                cosmologicalReconstructionCursor.next
            )

            interstellarReplicaNode.random = quantumEntanglementReplicaRegistry.get(
                cosmologicalReconstructionCursor.random
            )

            cosmologicalReconstructionCursor = cosmologicalReconstructionCursor.next

        return quantumEntanglementReplicaRegistry[head]